### PR TITLE
fixed(entities): Incorrect result on entities query

### LINKED ISSUE #63: Add test: Incorrect result on entities query (possibly special tautology case)
**Dataset:** https://github.com/Gapminder/ddf--pcbs--census/tree/master
**Query:**
```
{
  "language": "en",
  "from": "entities",
  "animatable": "year",
  "select": {
    "key": [
      "region"
    ],
    "value": [
      "name"
    ]
  },
  "where": {},
  "join": {},
  "order_by": [
    "rank"
  ]
}
```

**Expected response:**
```
0:{region: "1", name: "West Bank"}
1:{region: "2", name: "Gaza Strip"}
```

**Actual Response:**
```
0:{region: "pse", name: "Palestine"}
1:{region: "1", name: "West Bank"}
2:{region: "2", name: "Gaza Strip"}
3:{region: "1", name: "Jenin"}
4:{region: "1", name: "Tubas"}
5:{region: "1", name: "Tulkarm"}
6:{region: "1", name: "Qalqilya"}
7:{region: "1", name: "Salfit"}
8:{region: "1", name: "Nablus"}
9:{region: "1", name: "Ramallah & Al-Bireh"}
10:{region: "1", name: "Jerusalem"}
11:{region: "1", name: "Jericho"}
12:{region: "1", name: "Bethlehem"}
13:{region: "1", name: "Hebron"}
14:{region: "2", name: "North Gaza"}
15:{region: "2", name: "Gaza"}
16:{region: "2", name: "Deir Al - Balah"}
17:{region: "2", name: "Khanyunis"}
18:{region: "2", name: "Rafah"}
19:{region: "isr", name: "israel"}
20:{region: "not_stated", name: "Not Stated"}
21:{region: "other_countries", name: "Other Countries"}
22:{region: "1948_areas", name: "1948 areas"}
``` 

**Possible reason:**
Seeing `region` column in `ddf--entities--geo.csv` as key column containing entity id's, which it's not (as stated in datapackage.json resources table)

### PR DESCRIPTION
**Pull request by buchslava**
Closes #63

### CODE REVIEW

File: src/adapters/entity-adapter.ts [hunk 1]
```diff
@@ -131,6 +131,19 @@ export class EntityAdapter implements IDdfAdapter {
...
 
+  addIsClauseByKey(request) {
+    const isParentDomainPresent = entitySet => {
+      const domain = this.contentManager.domainHash[entitySet];
+
+      return includes(request.select.key, domain);
+    };
+    const extraIsClauses = request.select.key
+      .filter(concept => this.contentManager.conceptTypeHash[concept] === 'entity_set' && !isParentDomainPresent(concept))
+      .map(concept => ({$or: [{[`is--${concept}`]: 'TRUE'}, {[`is--${concept}`]: 'true'}]}));
```
Comment (jheeffer) [inline src/adapters/entity-adapter.ts]: why are you working with string boolean values?

**Comment by jheeffer** [comment]
How does this solve the core problem. What is your reasoning why adding `is--` checks always will give the right answer?

This `is--` checking shouldn't be necessary if you follow the key value pairs as defined in the resources section of datapackage.json. From them you can already see that `region` in the file is not the key but an entity property. So when a query with `key: region` comes in, the `region` column in the entities iss irrelevant. You're only interested in the key-column having entities that are `region` entities.

File: src/adapters/entity-adapter.ts (hunk 1 above)
Comment (buchslava) [reply src/adapters/entity-adapter.ts]: I this case I use values as-is. Is it wrong?

**Comment by buchslava** [comment]
My things are next:

1. First of all, in this case, we have only one file for processing - `ddf--entities--geo`. This fact was taken from `datapackage`
2. `ddf--entities--geo` is based on `geo` domain and contains different kinds of entity sets (including `region`)
3. We can use only one criterion to `region` extracting - `is--region` field
4. Because `where` section of our query is empty, we can't use it as-is without  any transformation
5. That's why I choose solution with additional clauses

Am I right? What flow do you propose for this case?

@angieskazka

File: src/adapters/entity-adapter.ts (hunk 1 above)
Comment (jheeffer) [reply src/adapters/entity-adapter.ts]: Does the reader do string comparison with the `is--` fields in the data? Or does it parse the queries `is--` fields to boolean?

**Comment by jheeffer** [comment]
It fixed this issue. But my point is that your fix seems to think there's ambiguity in the query. You try to solve that ambiguity by adding `is--` predicates to the `where` clause.
The thing is that there is no ambuigity. You don't need extra `is--` predicates if your further processing is correct. 

Your logic should:
 1. Get all resources for the key value pair
 2. In some order per resource:
    - Remove unqueried columns
    - Remove unqueried rows (using both the `select.key` clause and `where` clause)
        - `select.key` might contain entity sets in which case only rows with entities belonging to the set should be returned. Be sure to only check the primaryKey columns of the resource! This is probably where it goes wrong in current query logic, since it looks at the `region` column even though it's not in the primary key of the `ddf--entities--geo` resource.
        - `where` is normal where-clause filtering
    - Join resources to result

You don't need to normalize the query.

Furthermore, adding `is--` predicates might work for this solution but the core problem stays. This core problem might show later again. 

For example, what if it is a datapoint?
| geo | year | region |
| --- | --- | --- |
| 1 | 2015 |  |
| 1022 | 2015 | 1 | 

Adding `is--` predicates will not work when querying this table because you'd need a join clause to correctly add them.

**Comment by jheeffer** [comment]
Maybe it helps to look at my DDFcsv reader implementation. It needs a bit of clean up but the essence is right I believe.
https://github.com/jheeffer/ddfcsv-reader/blob/master/ddfcsv.js

**Comment by korel-san** [comment]
@jheeffer Hi! I've got your idea, and I know why @buchslava made his changes. I'm able to describe the main issue and provide you with all needed proves of our solution. But we need to finish our iteration today and start planning next iteration. So could you wait until we finish it for giving all explanations?

**Comment by buchslava** [comment]
@jheeffer @korel-san also: now I not sure that expected result in https://github.com/vizabi/vizabi-ddfcsv-reader/issues/63 is correct (I think the actual result is correct). @jheeffer please describe your logic `Your response logic should:...` on real example step by step (pcbs)

**Comment by jheeffer** [comment]
@korel-san sure, come back to it when you have time
@buchslava expected result is definitely correct

So let's take the query from #63:
```
{
  "select": {
    "key": [
      "region"
    ],
    "value": [
      "name"
    ]
  },
  "from": "entities"
```
So, it's asking for the key value pair `<region>,name` (`<key>,value`) from the pcbs dataset.

Now your response logic should:
1. Get the resources that contain this key value pair from datapackage.json ddfSchema. 
Result: resource `ddf--entities--geo`
datapackage.json: https://github.com/Gapminder/ddf--pcbs--census/blob/master/datapackage.json#L1048-L1056
jasper's ddfcsv reader: https://github.com/jheeffer/ddfcsv-reader/blob/master/ddfcsv.js#L63
2. Get the resource objects for these resources from datapackage.json in the resources array.
Result: https://github.com/Gapminder/ddf--pcbs--census/blob/master/datapackage.json#L682-L710
3. Load the csv file mentioned in this resource object:
csv file: https://github.com/Gapminder/ddf--pcbs--census/blob/master/datapackage.json#L683 and result: https://github.com/jheeffer/ddfcsv-reader/blob/master/ddfcsv.js#L63
ddfcsv reader: https://github.com/jheeffer/ddfcsv-reader/blob/master/ddfcsv.js#L64
4. Remove the unqueried columns from the resource (i.e. take the projection of the resource). Only value projection matters. Keys should always be returned.
`ddf--entities--geo` resource looks like:

| geo | name | is--country | is--region | is--governorate | region | country |
| --- | --- | --- | --- | --- | --- | --- |
| pse | Palestine | TRUE | FALSE | FALSE | | |
| 1 | West Bank | FALSE | TRUE | FALSE | | pse |
| 2 | Gaza Strip | FALSE | TRUE | FALSE | | pse |
| 1101 | Jenin | FALSE | FALSE | TRUE | 1 | pse |
| ... | ... | ... | ... | ... | ... | ... | 

Resulting projection of `{ value: ["name"] }` looks like:

| geo | name |
| --- | --- |
| pse | Palestine |
| 1 | West Bank | 
| 2 | Gaza Strip |
| 1101 | Jenin | 
| ... | ... | 

NOTE: `geo` in the 2nd table is NOT the `region` column in the first. It is the `geo` column with renamed header. The `region` column in the first is removed.

5. Rename the entity columns in the key to the queried key-columns get a correct response:

| region | name |
| --- | --- |
| pse | Palestine |
| 1 | West Bank | 
| 2 | Gaza Strip |
| 1101 | Jenin | 
| ... | ... | 

In my case I do renaming and projection at the same time. This might be improved to put both in separate operations.
ddfcsv-reader: https://github.com/jheeffer/ddfcsv-reader/blob/master/ddfcsv.js#L67 & https://github.com/jheeffer/ddfcsv-reader/blob/master/ddfcsv.js#L270

6. Remove rows that (a) do not satisfy the entity concepts in the queried key or (b) don't satisfy the `where` clause. In this case, (a) means removing rows where `region` is not in the region set. This is very similar to adding the `is--region: true` clause to the original query. However, you don't need to do this if the logic described here is implemented correctly. 
You can do this either by 
    1. joining `is--<entity_set>` on the resource and then filtering by `is--<entity_set>: { $eq: true }` 
    2. getting a list of entities belonging to the set and filtering with a `$in: [...]` filter

    In my code I do the 2nd. I don't know if there's big differences between the two methods. They seem two ways of doing the same. In your code you're doing the 1st. But in a "normalization" stage.
    code: https://github.com/jheeffer/ddfcsv-reader/blob/master/ddfcsv.js#L73-L75

So for this case I add merge `where` (empty) with the filter: `{ region: { $in: ["1", "2"] } }` and apply it resulting in:

| region | name |
| --- | --- |
| 1 | West Bank | 
| 2 | Gaza Strip |

That's the correct result.

**conclusion**
So, YES, you're right that adding `is--region: true` delivers correct results on entity queries. BUT:
1. On datapoint queries you need more logic as you need to join `is--region` on a part of the datapoint key.
2. It seems you already have logic in ddfcsv reader to solve entity set queries. However, it is not functioning well enough to correctly respond to the query in #63. Instead of fixing by putting a layer of normalization on top, it's better to go in the entity set logic and fix it there. 
My example only has one logic route for entity set queries (get members and add `$in: [..]` filter) and that is enough for all cases (as far as I can tell). You can use `is--<entity_set>` filter instead. That's fine, but there should be just one piece of logic to handle entity set row selection.

**Pull request closed without merging**